modules/acm/src/main/java/edu/gemini/epics/acm/CaTaskControlRecord.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2016-2018 Association of Universities for Research in Astronomy, Inc. (AURA)
+ * For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+package edu.gemini.epics.acm;
+
+import edu.gemini.epics.*;
+import edu.gemini.epics.api.ChannelListener;
+import edu.gemini.epics.impl.EpicsReaderImpl;
+import edu.gemini.epics.impl.EpicsWriterImpl;
+import gov.aps.jca.CAException;
+import gov.aps.jca.TimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class CaTaskControlRecord {
+    private static final Logger LOG = LoggerFactory.getLogger(CaApplyRecord.class.getName());
+
+    private static final String DIR_SUFFIX = ".DIR";
+    private static final String VAL_SUFFIX = ".VAL";
+    private static final String MSG_SUFFIX = ".MESS";
+    private static final String STATE_SUFFIX = ".BUSY";
+
+    private final String epicsName;
+
+    private EpicsReader epicsReader;
+    private EpicsWriter epicsWriter;
+    private ReadWriteClientEpicsChannel<CadDirective> dir;
+    private ReadOnlyClientEpicsChannel<Integer> val;
+    private ReadOnlyClientEpicsChannel<String> mess;
+    private ReadOnlyClientEpicsChannel<TaskControlState> busy;
+
+    private ChannelListener<Integer> valListener;
+    private ChannelListener<TaskControlState> busyListener;
+
+    CaTaskControlRecord(String epicsName, EpicsService epicsService) {
+        this.epicsName = epicsName;
+
+        epicsReader = new EpicsReaderImpl(epicsService);
+        epicsWriter = new EpicsWriterImpl(epicsService);
+
+        updateChannels();
+    }
+
+    synchronized void updateChannels() {
+        if (dir == null) {
+            try {
+                dir = epicsWriter.getEnumChannel(epicsName + DIR_SUFFIX, CadDirective.class);
+            } catch (Throwable e) {
+                LOG.warn(e.getMessage());
+            }
+        }
+        if (val == null) {
+            try {
+                val = epicsReader.getIntegerChannel(epicsName + VAL_SUFFIX);
+                if (valListener != null) {
+                    val.registerListener(valListener);
+                }
+            } catch (Throwable e) {
+                LOG.warn(e.getMessage());
+            }
+        }
+        if (mess == null) {
+            try {
+                mess = epicsReader.getStringChannel(epicsName + MSG_SUFFIX);
+            } catch (Throwable e) {
+                LOG.warn(e.getMessage());
+            }
+        }
+        if (busy == null) {
+            try {
+                busy = epicsReader.getEnumChannel(epicsName + STATE_SUFFIX, TaskControlState.class);
+                if (busyListener != null) {
+                    busy.registerListener(busyListener);
+                }
+            } catch (Throwable e) {
+                LOG.warn(e.getMessage());
+            }
+        }
+
+    }
+
+    synchronized void unbind() {
+        assert(epicsReader!=null);
```

Comment:
Do we really want `assert` here? I don't think they're intended for production code.

I think it might be better to use:

```
Objects.requireNonNull(epicsReader);
```